lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,377 @@
+import os
+import time
+from collections import OrderedDict
+
+
+permdict = {
+    "0": "---",
+    "1": "--x",
+    "2": "-w-",
+    "3": "-wx",
+    "4": "r--",
+    "5": "r-x",
+    "6": "rw-",
+    "7": "rwx",
+}
+
+
+def filepermissions(mode):
+    fperm = ""
+    modebits = oct(mode)[-3:]
+    for bits in modebits:
+        if bits in permdict:
+            fperm += permdict[bits]
+        else:
+            raise ValueError(
+                f"Could not find key '{bits}' in file permissions dictionary"
+            )
+
+    return fperm
+
+
+def header(fil, status, timestamp, hierarchy, path):
+    if status == "start":
+        fil.write(f"\n\n{status}-{timestamp}: {hierarchy} hierarchy: {path}\n")
+    else:
+        fil.write(f"\n{status}-{timestamp}: {hierarchy} hierarchy: {path}\n")
+
+
+def dump_check(fil, hierarchy_check, controller, key):
+    if controller in hierarchy_check[key]._dict:
+        fil.write(f"\t{hierarchy_check[key].get_msg()}\n")
+        output_format(fil, hierarchy_check[key]._dict[controller])
+    return 1
+
+
+def output_format(fil, controller_list):
```

Comment:
The `output_format` module method also seems like it could be a method of the `Hierarchy` base class.  Can you see how that might work, and reconcile it with the `output_format` of the other sub-classes of `Hierarchy`?

And, please add a method doc string, too.